internal/state/store.go

```diff
@@ -3,83 +3,280 @@ package state
 import (
 	"fmt"
 
-	v1 "k8s.io/api/core/v1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/relationship"
 )
 
-// store contains the resources that represent the state of the Gateway.
-type store struct {
-	gc         *v1beta1.GatewayClass
-	gateways   map[types.NamespacedName]*v1beta1.Gateway
-	httpRoutes map[types.NamespacedName]*v1beta1.HTTPRoute
-	services   map[types.NamespacedName]*v1.Service
-
-	// changed tells if the store is changed.
-	// The store is considered changed if:
-	// (1) Any of its resources was deleted.
-	// (2) A new resource was upserted.
-	// (3) An existing resource with the updated Generation was upserted.
+// Updater updates the cluster state.
+type Updater interface {
+	Upsert(obj client.Object)
+	Delete(objType client.Object, nsname types.NamespacedName)
+}
+
+// objectStore is a store of client.Object
+type objectStore interface {
+	get(nsname types.NamespacedName) (client.Object, bool)
+	upsert(obj client.Object)
+	delete(nsname types.NamespacedName)
+}
+
+// objectStoreMapAdapter wraps maps of types.NamespacedName to Kubernetes resources
+// (e.g. map[types.NamespacedName]*v1beta1.Gateway) so that they can be used through objectStore interface.
+type objectStoreMapAdapter[T client.Object] struct {
+	objects map[types.NamespacedName]T
+}
+
+func newObjectStoreMapAdapter[T client.Object](objects map[types.NamespacedName]T) *objectStoreMapAdapter[T] {
+	return &objectStoreMapAdapter[T]{
+		objects: objects,
+	}
+}
+
+func (m *objectStoreMapAdapter[T]) get(nsname types.NamespacedName) (client.Object, bool) {
+	obj, exist := m.objects[nsname]
+	return obj, exist
+}
+
+func (m *objectStoreMapAdapter[T]) upsert(obj client.Object) {
+	t, ok := obj.(T)
+	if !ok {
+		panic(fmt.Errorf("obj type mismatch: got %T, expected %T", obj, t))
+	}
+	m.objects[client.ObjectKeyFromObject(obj)] = t
+}
+
+func (m *objectStoreMapAdapter[T]) delete(nsname types.NamespacedName) {
+	delete(m.objects, nsname)
+}
+
+type gvkList []schema.GroupVersionKind
+
+func (list gvkList) contains(gvk schema.GroupVersionKind) bool {
+	for _, g := range list {
+		if gvk == g {
+			return true
+		}
+	}
+
+	return false
+}
+
+type multiObjectStore struct {
+	stores     map[schema.GroupVersionKind]objectStore
+	extractGVK extractGVKFunc
+}
+
+func newMultiObjectStore(
+	stores map[schema.GroupVersionKind]objectStore,
+	extractGVK extractGVKFunc,
```

Comment:
We inject this function because extraction depends on the scheme, which we inject in `NewChangeProcessorImpl`. The scheme needs to be injected because we create it in `manager.go`.

Note that we build this function in  `NewChangeProcessorImpl` based on the injected scheme:
```go
	extractGVK := func(obj client.Object) schema.GroupVersionKind {
		gvk, err := apiutil.GVKForObject(obj, cfg.Scheme)
		if err != nil {
			panic(fmt.Errorf("failed to get GVK for object %T: %w", obj, err))
		}
		return gvk
	}
```
uses the scheme. So we inject the scheme in the ChangeProcessor.